### DISCUSSION TITLE
Directory flattening pros and cons

### DISCUSSION THREAD

**Discussion by dkocher** [encryption-alliance/unified-vault-format#17] [General]
Opening this thread on the pros and cons on directory flattening to be taken into account in the decision making to include such a feature in the new format. It has been discussed in #5 that _omitting_ directory flattening in the new format has some advantages:

- Easier implementation [^1] 
- Keeping file hierarchy allows to apply ACLs
- Restore from backup is possible because of deterministic directory tree

However, having directory flattening also has its benefits:

- Working with object storage, where recursive rename and delete operations are not possible, renaming a directory only requires to move two files (`.c9r` and `.c9r/dir.c9r` with Cryptomator Vault Format 7) giving an incredible performance advantage when renaming directories with lots of children. This allows to perform working with encrypted storage better than when stored in a non-obfuscated directory tree.
- Working with vaults on a local filesystem with path length limitations (Windows) 

Please chime in if you are aware of other use cases that are impacted.

[^1]: https://github.com/cryptomator/new-encryption-format/discussions/5#discussioncomment-5023575

**Comment by overheadhunter** [comment]
Regarding the path length limit, there may be other solutions (with their own problems :sweat_smile:):
* On Windows, using the `\\?\` prefix
* Using Base4K encoding, as paths are not limited to 256 bytes but rather 256 multi-byte chars
* Stripping the IV from the file name (which rules out SIV), storing a per-directory IV (which is nonce-reuse-safe because child node names are unique)

Furthermore, it is worth noting, that even with a flat structure we may hit the path length limit due to a deeply nested base directory at which the vault is stored.

---

That said, let's look at the rename/move issue. You're right, it is a performance advantage, but the even stronger argument here is atomicity. We should always value integrity over performance.

In the linked discussion, I mistakenly claimed that such a move would be non-atomic but as you pointed out for S3 and the like the opposite is the case: A "flat" rename is atomic, while a "hierarchical" one is not. Which raises the question:

What would happen if a non-atomic "hierarchical" rename is cancelled mid-flight (regardless of whether cleartext or encrypted)? If every single rename operation is atomic (but the whole batch isn't), the user can simply re-try the incomplete rename, can't he? The worst case scenario would be that the user doesn't notice the incomplete rename at first, continues working on the new path, and (when he eventually reattempts the incomplete rename) runs into conflicts due to some nodes already being recreated by some other process.

---

However, as you already mentioned, restoring from a backup is only possible on a hierarchical structure (at least if we want to restore a dir with all its subdirs). Or is it?

What if a flat structure works in a way where each subdir is an actual symlink? Could a restore algorithm that follows symlinks work on this structure? If so, would this mean that we don't support symlinks pointing elsewhere?

```
vault/
├─ root/
│  ├─ ☞ dir1
│  ├─ file1
├─ dir1/
│  ├─ ☞ dir2
│  ├─ file2
│  ├─ file3
├─ dir2/
│  ├─ file4
```

**Reply by overheadhunter** [reply]
Some cloud and/or sync clients may not support syncing symlinks, though.

**Reply by overheadhunter** [reply]
@ncw Do you have an opinion on the proposed symlink-based solution (below the `<hr>` on my post above)? Do sync apps typically follow symlinks?

**Comment by overheadhunter** [comment]
Pro and con at the same time:

Making some folder offline-available does not apply to subfolders recursively.
